### PR TITLE
Implement eth_subscribe newHeads

### PR DESCRIPTION
**Pull request by bbondy**
This adds support for a new `eth_subscribe` method that can be used in Ethereum provider requests. 
See a description of this here: https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/subscription-methods/eth_subscribe

This implementation only supports `newHeads` for now.
There's a follow up issue in https://github.com/brave/brave-browser/issues/27283 for supporting the `"logs"` param.
These params are not currently supported by MM at the time of this writing, so they are probably not used in the wild and we can sit on the implementation for now: `"newPendingTransactions"` and `"syncing""`.

As an alternative to this implementation, I considered first adding websocket support and then using Infura, but then eth_subscribe wouldn't be supported for user's HTTPS configured RPC endpoints.  So I'm starting here with querying for block updates via the block tracker.

If approved I will do another pull request for wallet-docs.

Resolves https://github.com/brave/brave-browser/issues/21350

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Clone https://github.com/bbondy/eth-manual-tests and start a server as per the readme
- Use `eth_subscribe` to listen for new block head. You should see a console message in the developer console matching the same subscription ID every 20 seconds.  You can subscribe multiple times, each time will give a new subscription ID.
- Use `eth_unsubscribe` to stop listening to an event by putting the subscription ID from the `eth_subscribe` call.  It should stop ending events and return true to the console when you unsubscribe successfully from something.

### CODE REVIEW

File: components/brave_wallet/browser/ethereum_provider_impl.cc [hunk 1]
```diff
@@ -569,6 +574,43 @@ void EthereumProviderImpl::RecoverAddress(const std::string& message,
...
+    std::move(callback).Run(std::move(id), std::move(formed_response), reject,
+                            "", false);
+    return;
+  }
+  std::vector<uint8_t> bytes(16);
+  crypto::RandBytes(&bytes.front(), bytes.size());
+  std::string hex_bytes = ToHex(bytes);
+  eth_subscriptions_.push_back(hex_bytes);
+  if (eth_subscriptions_.size() == 1)
+    eth_block_tracker_.Start(base::Seconds(10));
```
Comment (darkdh) [inline components/brave_wallet/browser/ethereum_provider_impl.cc:594]: any reason that this is shorter than `kBlockTrackerDefaultTimeInSeconds` (20)?

File: browser/brave_wallet/ethereum_provider_impl_unittest.cc [hunk 2]
```diff
@@ -1964,6 +1986,62 @@ TEST_F(EthereumProviderImplUnitTest, AccountsChangedEvent) {
...
+  EXPECT_TRUE(response.second.is_string());
+  browser_task_environment_.FastForwardBy(
+      base::Seconds(kBlockTrackerDefaultTimeInSeconds));
+  EXPECT_TRUE(observer_->MessageEventFired());
+  base::Value rv = observer_->GetLastMessage();
+  ASSERT_TRUE(rv.is_dict());
+  auto& dict = rv.GetDict();
+  std::string* difficulty = dict.FindString("difficulty");
+  ASSERT_TRUE(difficulty);
+  EXPECT_EQ(*difficulty, "0x1");
```
Comment (darkdh) [inline browser/brave_wallet/ethereum_provider_impl_unittest.cc:2042]: Can we also do `eth_unsubscribe` with non existent subscription id and check `EthBlockTracker` is still running and the last subscription id to check if `EthBlockTracker` stop running?

**Comment by darkdh** [comment]
Can we also update https://github.com/bbondy/eth-manual-tests and add test plan for QA using it?

File: browser/brave_wallet/ethereum_provider_impl_unittest.cc (hunk 2 above)
Comment (bbondy) [reply browser/brave_wallet/ethereum_provider_impl_unittest.cc:2043]: Fixed in 60e51cb1134d0b20bb67da65ffdbd2808d25b0be
I also added a test to make sure it doesn't stop when a 2nd subscription exists.

File: components/brave_wallet/browser/ethereum_provider_impl.cc (hunk 1 above)
Comment (bbondy) [reply components/brave_wallet/browser/ethereum_provider_impl.cc]: Fixed / rebased in e1d053ba7448d9544d3ccbf92b1ebf2ce299df38

**Comment by bbondy** [comment]
> Can we also update https://github.com/bbondy/eth-manual-tests and add test plan for QA using it?

Yep, I'll do that today.

**Comment by bbondy** [comment]
Brave Wallet docs PR: https://github.com/brave/brave-wallet-docs/pull/84

**Comment by bbondy** [comment]
`eth_unsubscribe` PR for eth-manual-tests https://github.com/bbondy/eth-manual-tests/pull/6

**Pull request merged**